src/Interpreters/HashJoin.cpp

```diff
@@ -188,55 +198,101 @@ HashJoin::HashJoin(std::shared_ptr<TableJoin> table_join_, const Block & right_s
     , right_sample_block(right_sample_block_)
     , log(&Poco::Logger::get("HashJoin"))
 {
-    LOG_DEBUG(log, "Right sample block: {}", right_sample_block.dumpStructure());
+    LOG_TRACE(log, "Right sample block: {}", right_sample_block.dumpStructure());
+    bool multiple_disjuncts = key_names_right.size() > 1;
 
-    table_join->splitAdditionalColumns(right_sample_block, right_table_keys, sample_block_with_columns_to_add);
-    required_right_keys = table_join->getRequiredRightKeys(right_table_keys, required_right_keys_sources);
+    if (multiple_disjuncts)
+    {
+        // required_right_keys_sources concept does not work well if multiple disjuncts
```

Comment:
The original idea is if join condition is ON T1.A = T2.A it is pointless to keep in memory both T1.A and T2.A. If a row exists in resultset, we can use value of T1.A instead of T2.A.
It does not work if we have OR (in T1.A = T2.A or T1.B = T2.B it is not clear which equality is true) and we need all columns.
The original behavior is preserved because the goal was zero cost enhancement although I _think_ that from functional point of view it is Ok to get rid of this optimization completely. 
Reconfirm if I should test if it is really true.